rectools/visuals/metrics_app.py

```diff
@@ -85,62 +96,160 @@ def construct(
+        # Validate that all Models names are unique
+        if Columns.Split in models_metrics.columns:
+            models_metrics[Columns.Model] = models_metrics[Columns.Model].astype(str).str.replace(" ", "_")
+            models_names_comb = models_metrics[Columns.Model] + models_metrics[Columns.Split].astype(str)
+            if models_names_comb.nunique() != len(models_names_comb):
+                raise ValueError("Each `Model` value in the `metrics_data` DataFrame must be unique")
+        else:
+            models_metrics[Columns.Model] = models_metrics[Columns.Model].astype(str).str.replace(" ", "_")
+            if models_metrics[Columns.Model].nunique() != len(models_metrics[Columns.Model]):
+                raise ValueError("Each `Model` value in the `metrics_data` DataFrame must be unique")
+
+    @staticmethod
+    def _validate_models_metadata(models_metadata: pd.DataFrame) -> None:
+        if not isinstance(models_metadata, pd.DataFrame):
+            raise ValueError("Incorrect input type. `models_metadata` should be a DataFrame")
+        if Columns.Model not in models_metadata.columns:
+            raise KeyError("Missing `Model`` column in `models_metadata` DataFrame")
+        if models_metadata[Columns.Model].nunique() != len(models_metadata):
+            raise ValueError("`Model` values of `models_metadata` should be unique`")
+        if models_metadata[Columns.Model].isnull().any():
+            raise ValueError("Found NaN values in `Model` column")
 
     @lru_cache
-    def _make_chart_data(self, fold_number: int) -> pd.DataFrame:  # pragma: no cover
-        return self.models_metrics[self.models_metrics[Columns.Split] == fold_number].drop(columns=Columns.Split)
+    def _make_chart_data_fold(self, fold_number: int) -> pd.DataFrame:
+        return self.data[self.data[Columns.Split] == fold_number].reset_index(drop=True)
 
     @lru_cache
     def _make_chart_data_avg(self) -> pd.DataFrame:
-        return self.models_metrics.drop(columns=Columns.Split).groupby(Columns.Model, sort=False).mean().reset_index()
+        metric_data_columns = [Columns.Model] + self.metric_names
+        meta_data_columns = [Columns.Model] + self.meta_names
+        # separate metric data because meta data could contain strings
+        metrics_data = self.data[metric_data_columns].groupby([Columns.Model], sort=False).mean().reset_index()
+        meta_data = self.data[meta_data_columns].drop_duplicates()
+        return metrics_data.merge(meta_data, on=Columns.Model, how="left").reset_index(drop=True)
+
+    def _create_chart(
+        self,
+        data: pd.DataFrame,
+        metric_x: str,
+        metric_y: str,
+        color: str,
+        legend_title: str,
+    ) -> go.Figure:
+        scatter_kwargs = {
+            "width": WIDGET_WIDTH,
+            "height": WIDGET_HEIGHT,
+        }
+        scatter_kwargs.update(self.scatter_kwargs)
+
+        data = data.sort_values(by=color, ascending=False)
```

Comment:
How about changing it go `ascending=True`?
And I wrote to you about recolouring `nan` to grey separately

Reply:
Done, but current recolouring is black